### ISSUE TITLE
Temporal regularization experiment with traffic usage data

### ISSUE THREAD

**Issue by deploy-soon** [deploy-soon/TemporalMF#3]
Some experiment results which contain validation loss among several loss functions.
Traffic usage data is from [Multivariate Time series Data sets](https://github.com/laiguokun/multivariate-time-series-data).

**Traffic Usage**
The raw data is in http://pems.dot.ca.gov. The data in this repo is a collection of 48 months (2015-2016) hourly data from the California Department of Transportation. The data describes the road occupancy rates (between 0 and 1) measured by different sensors on San Francisco Bay area freeways.

shape: (17544, 862)

**Comment by deploy-soon** [comment]
Base Matrix Factorization
- code: [`10d386`](https://github.com/deploy-soon/TemporalMF/commit/10d386605e47fb931e67027e7b48c5058d2ac242)
- loss function
![image](https://user-images.githubusercontent.com/44163510/75543102-71298f80-5a64-11ea-8e07-8c311f1d8b3f.png)
- Adam optimizer(learning rate=0.0005)
- train ratio: 0.8
- metric:
![image](https://user-images.githubusercontent.com/44163510/75543454-29573800-5a65-11ea-81f1-fe1e96da7eaf.png)
- result

|**Factors** | 20     | 30     | 40     | 50     |
|---------|--------|--------|--------|--------|
| **NRMSE**   | 0.6845 | 0.7784 | 0.8951 | 1.022 |

**Comment by deploy-soon** [comment]
L2 Reg Matrix Factorization
- code: [`10d386`](https://github.com/deploy-soon/TemporalMF/commit/10d386605e47fb931e67027e7b48c5058d2ac242)
- loss function
![image](https://user-images.githubusercontent.com/44163510/75551705-1ef26980-5a78-11ea-9d07-3e4c9404aab9.png)
- Adam optimizer(learning rate=0.005)
- train ratio: 0.8
- metric:
![image](https://user-images.githubusercontent.com/44163510/75543454-29573800-5a65-11ea-81f1-fe1e96da7eaf.png)
- result

| Factors      | 20     | 30     | 40     | 50 |
|--------------|--------|--------|--------|----|
| lambda=5.0   |  1.381 | 1.381  | 1.381  | 1.381  |
| lambda=0.5   | 1.381  | 1.382 | 1.381 |  1.381  |
| lambda=0.05  | 1.097  | 1.096 | 1.098  | 1.098   |
| lambda=0.005 | *0.5997 | 0.6148 | 0.6317 |  0.6484  |

**Comment by deploy-soon** [comment]
Temporal L2 Reg Matrix Factorization
- code: [`10d386`](https://github.com/deploy-soon/TemporalMF/commit/10d386605e47fb931e67027e7b48c5058d2ac242)
- loss function
![image](https://user-images.githubusercontent.com/44163510/75846026-4d789780-5e1e-11ea-8cf7-fb51e509ffa6.png)
- Adam optimizer(learning rate=0.0005)
- train ratio: 0.8
- metric:
![image](https://user-images.githubusercontent.com/44163510/75543454-29573800-5a65-11ea-81f1-fe1e96da7eaf.png)
**result**
- temporal lambda = lambda / 10.0

| Factors      | 20     | 30     | 40     | 50 |
|--------------|--------|--------|--------|----|
| lambda=5.0   | 1.381  | 1.381  | 1.381  |  1.381  |
| lambda=0.5   | 1.382 | 1.381 | 1.382 |  1.381  |
| lambda=0.05  | 0.9069 | 0.9053 | 0.9032 |    |
| lambda=0.005 | 0.5974 | 0.6184 | 0.6479  |    |

  - temporal lambda = lambda / 1.0

| Factors      | 20     | 30     | 40     | 50 |
|--------------|--------|--------|--------|----|
| lambda=5.0   | 1.381  | 1.382  |  1.381 |  1.381  |
| lambda=0.5   | 1.381 | 1.381 | 1.382 |    |
| lambda=0.05  | 0.9135 | 0.9123 | 0.9143 |    |
| lambda=0.005 | 0.5801 | 0.5964 | 0.6132 |    |


  - temporal lambda = lambda / 0.1

| Factors      | 20     | 30     | 40     | 50 |
|--------------|--------|--------|--------|----|
| lambda=5.0   | 1.382  | 1.382  | 1.381  |  1.382  |
| lambda=0.5   | 1.381 |1.381  | 1.381 |    |
| lambda=0.05  | 0.9599 | 0.9566 | 0.9616 |    |
| lambda=0.005 | 0.5611 | 0.5624 | 0.5658 |    |


  - temporal lambda = lambda / 0.01

| Factors      | 20     | 30     | 40     | 50 |
|--------------|--------|--------|--------|----|
| lambda=5.0   | 1.382  | 1.381  |  1.382 |  1.382  |
| lambda=0.5   | 1.381 | 1.382 | 1.382 |    |
| lambda=0.05  | 1.071 | 1.381 | 1.0715 |    |
| lambda=0.005 | 0.6634 | 0.5904 | 0.5888 |    |

**Comment by deploy-soon** [comment]
Test loss for several recurrent layers and dimensions just LSTM.

- dropout rate: 0.3
- lambda_x = 5.0
- lags: 100

|LSTM|factors=20|factors=40|factors=80|
|---|---|---|---|
|n_layers=1 hid_dim=128|0.6393 447500|0.6179 828440|0.6229 1590320|
|n_layers=1 hid_dim=256|0.6371 657932|0.6189 1051672|0.6143 1839152|
|n_layers=2 hid_dim=128|0.6313 579596|0.6182 960536|0.6082 1722416|
|n_layers=2 hid_dim=256|0.6316 1184268|0.6206 1578008|0.6122 2365488|
